### PR TITLE
fix pserver sub-blocks

### LINKED ISSUE #11429: Error when transpile program with piecewise_decay to distributed program
When `piecewise_decay` is defined the pserver side program will have a wrong `conditional_block` refer to a block id that doesn't exist on the pserver program.

This error was first met by @kolinwei 

```python
optimizer = fluid.optimizer.Momentum(
            learning_rate=fluid.layers.piecewise_decay(
                boundaries=bd, values=lr),
            momentum=0.9,
            regularization=fluid.regularizer.L2Decay(1e-4))
```

### PR DESCRIPTION
**Pull request by Yancey0623**
Fixed #11429

### CODE REVIEW

**Comment by velconia** [comment]
Should we remove these sub blocks from blk_list in the RunAsyncLoop method of listen_and_serv_op, like what we do in the RunSyncLoop method?

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 1]
```diff
@@ -479,16 +479,17 @@ def __op_have_grad_input__(op):
                     return varname
             return ""
 
-        def __clone_lr_op_sub_block__(op, program, new_block):
+        def __clone_lr_op_sub_block__(op, program, new_block, skip_sub_blks):
             if not op.has_attr('sub_block'):
-                return
+                return -1
```
Comment (gongweibao) [inline python/paddle/fluid/transpiler/distribute_transpiler.py]: Who uses the return value?

**Comment by Yancey0623** [comment]
> Should we remove these sub blocks from blk_list in the RunAsyncLoop method of listen_and_serv_op, like what we do in the RunSyncLoop method?

May not, RunAsyncLoop use `grad_to_block_id` to determinate which optimize block to be execute by gradient var, and also don't support lr decay in async update.

File: paddle/fluid/operators/listen_and_serv_op.cc [hunk 2]
```diff
@@ -101,13 +101,16 @@ void ListenAndServOp::RunSyncLoop(
...
   size_t num_blocks = program->Size();
+  auto skip_sub_blks = Attr<std::vector<int>>("skip_sub_blks");
   PADDLE_ENFORCE_GE(num_blocks, 2,
                     "server program should have at least 2 blocks");
 
   std::vector<int> optimize_block_id_list;
   for (int blkid = 1; blkid < num_blocks; ++blkid) {
     if (std::find(prefetch_block_id_list.begin(), prefetch_block_id_list.end(),
-                  blkid) == prefetch_block_id_list.end()) {
+                  blkid) == prefetch_block_id_list.end() &&
```
Comment (gongweibao) [inline paddle/fluid/operators/listen_and_serv_op.cc]: Is there a way to make it beautiful?

File: paddle/fluid/operators/listen_and_serv_op.cc [hunk 3]
```diff
@@ -101,13 +101,16 @@ void ListenAndServOp::RunSyncLoop(
     framework::Scope *recv_scope,
     const std::vector<int> &prefetch_block_id_list) const {
   size_t num_blocks = program->Size();
+  auto skip_sub_blks = Attr<std::vector<int>>("skip_sub_blks");
```
Comment (typhoonzero) [inline paddle/fluid/operators/listen_and_serv_op.cc]: Record only the start block id and end block id for `listen_and_serv` op to run parallelly. Sub-blocks append to the end to make the program in a strict order.

Comment (Yancey0623) [reply paddle/fluid/operators/listen_and_serv_op.cc]: Another way is that we record the parallel block ids, the benefit is don't need `insert` block to the specified idx, just append block to pserver_program and record the optimize block id.

Comment (typhoonzero) [reply paddle/fluid/operators/listen_and_serv_op.cc]: Reuse that deprecated `optimize_block` attribute could be cool.

Comment (Yancey0623) [reply paddle/fluid/operators/listen_and_serv_op.cc]: Will do that.

File: python/paddle/fluid/transpiler/distribute_transpiler.py (hunk 1 above)
Comment (Yancey0623) [reply python/paddle/fluid/transpiler/distribute_transpiler.py]: Done.

File: paddle/fluid/operators/listen_and_serv_op.cc (hunk 2 above)
Comment (Yancey0623) [reply paddle/fluid/operators/listen_and_serv_op.cc]: Done, please follow the comment: https://github.com/PaddlePaddle/Paddle/pull/11585#pullrequestreview-130377209

**Review by typhoonzero — DISMISSED** [review]
LGTM, thanks this should work.

**Review by typhoonzero — APPROVED** [review]
